src/gt4py/next/ffront/foast_passes/type_deduction.py

```diff
@@ -891,6 +890,7 @@ def _visit_astype(self, node: foast.Call, **kwargs: Any) -> foast.Call:
                 primitive_type, getattr(ts.ScalarKind, new_type.id.upper())
             ),
         )
+        assert isinstance(return_type, (ts.TupleType, ts.ScalarType, ts.FieldType))
```

Comment:
Just a question, why is it needed to replace the static type check by a runtime check here?

Reply:
`apply_to_primitive_constituents` was not typed, but the return is not precise enough for how it's used here. Didn't check if it should be loosened here.